storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operation/get/GetObjectsStub.java

```diff
@@ -0,0 +1,226 @@
+
+        query = mock(Query.class);
+        when(query.table()).thenReturn("test_table");
+
+        rawQuery = mock(RawQuery.class);
+        when(rawQuery.observesTables()).thenReturn(singleton("test_table"));
+
+        getResolver = mock(GetResolver.class);
+        cursor = mock(Cursor.class);
+
+        items = new ArrayList<TestItem>();
+        items.add(new TestItem());
+        items.add(new TestItem());
+        items.add(new TestItem());
+
+        when(cursor.getCount())
+                .thenReturn(items.size());
+
+        when(cursor.moveToNext()).thenAnswer(new Answer<Boolean>() {
+            int invocationsCount = 0;
+
+            @Override
+            public Boolean answer(InvocationOnMock invocation) throws Throwable {
+                return invocationsCount++ < items.size();
+            }
+        });
+
+        when(storIOSQLite.get())
+                .thenReturn(new PreparedGet.Builder(storIOSQLite));
+
+        when(storIOSQLite.observeChangesInTables(eq(singleton(query.table()))))
+                .thenReturn(Observable.<Changes>empty());
+
+        assertNotNull(rawQuery.observesTables());
+
+        when(storIOSQLite.observeChangesInTables(rawQuery.observesTables()))
+                .thenReturn(Observable.<Changes>empty());
+
+        when(getResolver.performGet(storIOSQLite, query))
+                .thenReturn(cursor);
+
+        when(getResolver.performGet(storIOSQLite, rawQuery))
+                .thenReturn(cursor);
+
+        when(getResolver.mapFromCursor(cursor))
+                .thenAnswer(new Answer<TestItem>() {
+                    int invocationsCount = 0;
+
+                    @Override
+                    public TestItem answer(InvocationOnMock invocation) throws Throwable {
+                        final TestItem testItem = items.get(invocationsCount);
+                        invocationsCount++;
+                        return testItem;
+                    }
+                });
+
+        typeMapping = mock(SQLiteTypeMapping.class);
+
+        if (withTypeMapping) {
+            when(internal.typeMapping(TestItem.class)).thenReturn(typeMapping);
+            when(typeMapping.getResolver()).thenReturn(getResolver);
+        }
+    }
+
+    @NonNull
+    static GetObjectsStub newInstanceWithoutTypeMapping() {
+        return new GetObjectsStub(false);
+    }
+
+    @NonNull
+    static GetObjectsStub newInstanceWithTypeMapping() {
+        return new GetObjectsStub(true);
+    }
+
+    void verifyQueryBehavior(@NonNull List<TestItem> actualList) {
+        // should be called once
+        verify(storIOSQLite).get();
+
+        // should be called only once
+        verify(getResolver).performGet(storIOSQLite, query);
+
+        // should be called same number of times as number of items
+        verify(getResolver, times(items.size())).mapFromCursor(cursor);
+
+        // should be called only once because of Performance!
+        verify(cursor).getCount();
```

Comment:
Yep, to create list of exact required size, also don't call 'count' more
than once.
20 июня 2015 г. 12:06 PM пользователь "Dmitrii Nikitin" <
notifications@github.com> написал:

> In
> storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operation/get/GetObjectsStub.java
> <https://github.com/pushtorefresh/storio/pull/410#discussion_r32883486>:
>
> > +    static GetObjectsStub newInstanceWithTypeMapping() {
> > +        return new GetObjectsStub(true);
> > +    }
> > +
> > +    void verifyQueryBehavior(@NonNull List<TestItem> actualList) {
> > +        // should be called once
> > +        verify(storIOSQLite).get();
> > +
> > +        // should be called only once
> > +        verify(getResolver).performGet(storIOSQLite, query);
> > +
> > +        // should be called same number of times as number of items
> > +        verify(getResolver, times(items.size())).mapFromCursor(cursor);
> > +
> > +        // should be called only once because of Performance!
> > +        verify(cursor).getCount();
>
> getCount because of performance?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pushtorefresh/storio/pull/410/files#r32883486>.
>